### ISSUE TITLE
Outstanding SendAsync call

### ISSUE THREAD

**Issue by poettlr** [Marfusios/websocket-client#18]
Hello, 

Unfortunately, I can't reproduce my problem but I sometimes get the following error: 
> Error: There is already one outstanding 'SendAsync' call for this WebSocket instance. ReceiveAsync and SendAsync can be called simultaneously, but at most one outstanding operation for each of them is allowed at the same time.

This seems to be also the cause of another error happening later during execution of my app: 
>Error: Cannot access a disposed object. Object name: 'System.Net.WebSockets.ClientWebSocket'.

Can you confirm, that the first error disposes of the ClientWebSocket and that will lead to the second error? 

I use the following snippet to send my messages (and never use SendInstant): 
`public void SendCommand(JObject command) {
                client.Send(command.ToString);
}`

Do you have any idea, why there might be an outstanding send? 

Thank you, 
Cheers,

**Comment by Marfusios** [comment]
Hello, 

thanks for reporting the issue. 
`Send()` will be fine, because it is solved by queue and processed one by one, but from the error info, it seems to be related to `ReceiveAsync` which couldn't be called simultaneously. 
Probably it could happen on intensive reconnections. I suspect missing thread locking at two places, which causes race condition with reconnection. 

`Disposed object` most likely connected with this base issue. 
I will investigate it further.

**Comment by poettlr** [comment]
Is there anything I can add on my side to detect that? 
Should I wrap the send call into a try/catch and then just try to recreate the connection?

**Comment by poettlr** [comment]
I want to provide extra information. My usage pattern of the Websocket.Client involves: 
* Using client.Send inside of a ReconnectCallback 
* Using client.Send inside of a MessageReceivedCallback 

Maybe that helps, and/or the issue is on my end of the code?

**Comment by Marfusios** [comment]
Hello @poettlr,

I did a few checks and was able to reproduce the issue with this code: 

```csharp
client.MessageReceived.Subscribe(msg =>
{
	counter++;
	Log.Information($"Message received: {msg}");

	if(counter % 10 == 0)
	{
		client.Reconnect();
		client.Reconnect();
		Task.Delay(1000).Wait();
		client.Reconnect();
		client.Reconnect();
	}
});
```

I've implemented synchronization between reconnection and sending, it should help. 
Please try latest version [3.1.32](https://www.nuget.org/packages/Websocket.Client/3.1.32)

**Comment by poettlr** [comment]
Awesome I will review and try the code on the office next week!
Thanks!

**Comment by Marfusios** [comment]
@poettlr did it help? I've implemented [version 4.0](https://github.com/Marfusios/websocket-client/releases/tag/v4.0), please try
